src/main/java/graphql/util/RecursionState.java

```diff
@@ -1,70 +1,97 @@
 package graphql.util;
 
+import graphql.Assert;
 import graphql.Internal;
 
 import java.util.ArrayDeque;
 import java.util.Collection;
 import java.util.Deque;
-import java.util.HashMap;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
-import java.util.stream.Collectors;
 
 import static graphql.Assert.assertNotNull;
 
 @Internal
-abstract class RecursionState<T> {
+public class RecursionState<T> {
 
-    private final Deque<TraverserContext<?>> delegate;
-    private final Map<T, Object> visitedMap = new ConcurrentHashMap<>();
+    private Object initialData;
 
-    public RecursionState() {
-        this(new ArrayDeque<>(32));
+    public enum Type {
+        STACK,
+        QUEUE
     }
 
-    public RecursionState(Deque<? super TraverserContext<T>> delegate) {
-        this.delegate = (Deque<TraverserContext<?>>) assertNotNull(delegate);
+    private final Deque<Object> state;
+    private final Set<T> visited = new LinkedHashSet<>();
+    private final Type type;
+
+
+    public enum Marker {
+        END_LIST
     }
 
-    public TraverserContext<T> peek() {
-        return (TraverserContext<T>) delegate.peek();
+    public RecursionState(Type type, Object initialData) {
+        this.initialData = initialData;
+        this.state = new ArrayDeque<>(32);
+        this.type = assertNotNull(type);
     }
 
-    public abstract TraverserContext<T> pop();
 
-    public abstract void pushAll(TraverserContext<T> o, Function<? super T, ? extends List<T>> getChildren);
+    public Object peek() {
+        return state.peek();
+    }
+
+
+    public Object pop() {
+        return this.state.pop();
+    }
+
+    public void pushAll(TraverserContext<T> o, Function<? super T, ? extends List<T>> getChildren) {
+        if (type == Type.QUEUE) {
+            getChildren.apply(o.thisNode()).iterator().forEachRemaining((e) -> this.state.add(newContext(e, o)));
+            this.state.add(Marker.END_LIST);
+            this.state.add(o);
+        } else if (type == Type.STACK) {
+            this.state.push(o);
+            this.state.push(Marker.END_LIST);
+            new ArrayDeque<>(getChildren.apply(o.thisNode())).descendingIterator().forEachRemaining((e) -> this.state.push(newContext(e, o)));
```

Comment:
This is partially my fault:)

getChildren() returns List<Node>.
Instead of `new ArrayDeque<>(getChildren(...)).descendingIterator()` I think it would be more efficient to do `getChildren().listIterator(size())` and wrap it into an Iterator that goes in the reverse direction. I'll submit PR for this